bazel/foreign_cc/BUILD

```diff
@@ -237,4 +253,46 @@ envoy_cmake_external(
         "//bazel:windows_x86_64": ["zlibstatic.lib"],
         "//conditions:default": ["libz.a"],
     }),
+    variants = [
+        {
+            "name": "zlib_ng",
+            "cache_entries": {
+                # zlib-ng build options. Reference: https://github.com/zlib-ng/zlib-ng#build-options.
+                "ZLIB_COMPAT": "on",
+                "ZLIB_ENABLE_TESTS": "off",
+
+                # Turning off WITH_OPTIM and WITH_NEW_STRATEGIES fixes checksum, but not sure if that is
+                # what we want.
+                "WITH_OPTIM": "off",
+                "WITH_NEW_STRATEGIES": "off",
+
+                # zlib-ng advance build options. Reference: https://github.com/zlib-ng/zlib-ng#advanced-build-options.
+                # Only allow aligned address.
+                "UNALIGNED_OK": "off",
+            },
+            "lib_source": "@com_github_zlib_ng_zlib_ng//:all",
+        },
+        {
+            "name": "zlib_ng_with_optimizations",
+            "cache_entries": {
+                # zlib-ng build options. Reference: https://github.com/zlib-ng/zlib-ng#build-options.
+                "ZLIB_COMPAT": "on",
+                "ZLIB_ENABLE_TESTS": "off",
+
+                # Warning: Turning WITH_OPTIM to "on" doesn't pass ZlibCompressorImplTest.CallingChecksum.
+                "WITH_OPTIM": "on",
+                # However turning off SSE4 fixes it.
+                "WITH_SSE4": "off",
+
+                # Warning: Turning WITH_NEW_STRATEGIES to "on" doesn't pass gzip compressor fuzz test.
+                # Turning this off means falling into NO_QUICK_STRATEGY route.
+                "WITH_NEW_STRATEGIES": "off",
+
+                # zlib-ng advance build options. Reference: https://github.com/zlib-ng/zlib-ng#advanced-build-options.
+                # Only allow aligned address.
+                "UNALIGNED_OK": "off",
```

Comment:
I tweaked the build options to be this way to make sure we can get through the required tests in CI for `--define ng-with-optimizations`. Feel free to tweak it further. cc. @rgs1

Reply:
LGTM -- thanks!